Mage.Sets/src/mage/cards/a/AgencyOutfitter.java

```diff
@@ -0,0 +1,127 @@
+        this.toughness = new MageInt(3);
+
+        // Flying
+        this.addAbility(FlyingAbility.getInstance());
+
+        // When Agency Outfitter enters the battlefield, you may search your graveyard, hand, and/or library for a card named Magnifying Glass and/or a card named Thinking Cap and put them onto the battlefield. If you search your library this way, shuffle.
+        this.addAbility(new EntersBattlefieldTriggeredAbility(new AgencyOutfitterEffect(), true));
+    }
+
+    private AgencyOutfitter(final AgencyOutfitter card) {
+        super(card);
+    }
+
+    @Override
+    public AgencyOutfitter copy() {
+        return new AgencyOutfitter(this);
+    }
+}
+
+//Based on Boonweaver Giant / Dark Supplicant / Gate to the Afterlife / Mishra, Artificer Prodigy
+class AgencyOutfitterEffect extends OneShotEffect {
+
+    AgencyOutfitterEffect() {
+        super(Outcome.UnboostCreature);
+        this.staticText = "you may search your graveyard, hand, and/or library for a card named Magnifying Glass and/or a card named Thinking Cap and put them onto the battlefield. If you search your library this way, shuffle.";
+    }
+
+    private AgencyOutfitterEffect(final AgencyOutfitterEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public AgencyOutfitterEffect copy() {
+        return new AgencyOutfitterEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+
+        //Uses AbstractMap.SimpleImmutableEntry as a generic Pair class
+        //Create a local lambda function to reduce code duplication
+        Function<String, AbstractMap.SimpleImmutableEntry<Card, Boolean>> searchEverywhereForCard = (name) -> {
+            FilterCard filter = new FilterCard(name);
+            filter.add(new NamePredicate(name));
+            Card card = null;
+            boolean searchedLibrary = false;
+
+            // Choose card from graveyard
+            if (controller.chooseUse(Outcome.Neutral, "Search your graveyard for "+name+"?", source, game)) {
+                TargetCardInYourGraveyard target = new TargetCardInYourGraveyard(filter);
+                if (controller.choose(Outcome.PutCardInPlay, controller.getGraveyard(), target, source, game)) {
+                    card = game.getCard(target.getFirstTarget());
+                }
+            }
+
+            // Choose card from your hand
+            if (card == null && controller.chooseUse(Outcome.Neutral, "Search your hand for "+name+"?", source, game)) {
+                TargetCardInHand target = new TargetCardInHand(filter);
+                if (controller.choose(Outcome.PutCardInPlay, controller.getHand(), target, source, game)) {
+                    card = game.getCard(target.getFirstTarget());
+                }
+            }
+
+            // Choose a card from your library
```

Comment:
@xenohedron I don’t remember — is multiple zones search doing by single choose dialog now (from whole cards list) instead multiple choices?

Reply:
The four cards I based this implementation on (Boonweaver Giant / Dark Supplicant / Gate to the Afterlife / Mishra, Artificer Prodigy) all use this pattern for their searches, though I added an option to not search the library instead of assuming that if a card wasn't found by that point that the library should be searched.

Reply:
I remember some refactors and discussions about multi zones searches by single list/dialog. If it works and has better UX for a player then must be used. Example: #11213

Reply:
It doesn't work as-is, but I could make a new Effect to support searching all three zones (and then have all five of these cards to use that effect).

Reply:
It seems there are still a bunch of existing cards that search one zone and then the next. I don't like that paradigm, with extra `chooseUse` and such, if it can be avoided. `PutCardFromHandOrGraveyardOntoBattlefieldEffect` is a recent streamlined example.

Searching library is a little different though. You can't otherwise just look at your library. You need the option to not search it and thereby not shuffle it. But I think it would suffice to have a single prompt for whether to include library in the search.

Then you should search for both targets simultaneously. See Alpine Houndmaster for an example, using `TargetCardWithDifferentNameInLibrary`.

So approach outline for new effect class:
- Add all cards matching filter in hand/graveyard to CardsImpl
- Ask whether to search library, if so, also add cards there matching filter
- Search those cards (also reveal full library window if it is part of the search, I guess) using new `TargetCardWithDifferentName` class extending `TargetCard` with existing logic

I'm pretty sure Boonweaver Giant will need to remain custom effect. But it's fine to allow only a single card name and still use the different name logic.

I recommend new issue to clarify scope, and specific PR dedicated to that scope. (I don't like including multiple card implementations in one PR unless they are clearly related. It complicates the review.)